internal/buff/write.go

```diff
@@ -0,0 +1,167 @@
+// This source file is part of the EdgeDB open source project.
+//
+// Copyright 2020-present EdgeDB Inc. and the EdgeDB authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package buff
+
+import (
+	"encoding/binary"
+	"io"
+
+	"github.com/edgedb/edgedb-go/types"
+)
+
+// Writer is a write buffer.
+type Writer struct {
+	alocatedMemory [256 * 1024]byte
+	buf            []byte
+
+	msgPos  int
+	bytePos []int
+}
+
+// NewWriter returns a new Writer.
+func NewWriter() *Writer {
+	w := &Writer{}
+	w.buf = w.alocatedMemory[:0]
+	return w
+}
+
+// Send writes buffered data to conn.
+func (w *Writer) Send(conn io.Writer) (err error) {
+	if w.msgPos != 0 {
+		panic("cannot send: the previous message is not finished")
+	}
+
+	if len(w.buf) == 0 {
+		panic("cannot send: no data")
+	}
+
+	var n int
+	n, err = conn.Write(w.buf)
+	if err != nil {
+		return err
+	}
+
+	if n != len(w.buf) {
+		panic("not all data was sent")
```

Comment:
Hm, why panic and not try sending the remaining data? Is writing always supposed to succeed and panicing is the recommended way to deal with it?

Reply:
No. Good catch. This is incomplete :confused:

Reply:
updated :heavy_check_mark: